### PR TITLE
837 new policies

### LINKED ISSUE #837: Changes to Kusk Policy
Proposal to change the API of the Kusk Policy

## Auth Policies 

Generally removing the `scheme` object and making it more relevant to the specific provider: 

**Cloudentity**
```yaml
# FROM 
x-kusk:
  auth:
    scheme: cloudentity
    auth-upstream:
      host:
        hostname: cloudentity-authorizer-standalone-authorizer.kusk-system
        port: 9004

# TO
x-kusk:
  auth:
    cloudentity:
      host:
        hostname: cloudentity-authorizer-standalone-authorizer.kusk-system
        port: 9004
```

**Basic auth:**

```yaml
# Basic -> Rename to custom-upstream
# FROM 
x-kusk:
  auth:
    scheme: basic
    path_prefix: /login # optional
    auth-upstream:
      host:
        hostname: example.com
        port: 80

# TO
x-kusk:
  auth:
    custom:
      host:
        hostname: example.com
        port: 80
```

**OAuth2**

```yaml
# OAuth2
# FROM 
x-kusk:
  auth:
    scheme: oauth2
    oauth2:
      token_endpoint: *TOKEN_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/oauth/token
      authorization_endpoint: *AUTHORIZATION_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/authorize
      credentials:
        client_id: *CLIENT_ID*
        client_secret: *CLIENT_SECRET*
      redirect_uri: /oauth2/callback
      redirect_path_matcher: /oauth2/callback
      signout_path: /oauth2/signout
      forward_bearer_token: true
      auth_scopes:
        - openid

# TO 
x-kusk:
  auth:
    oauth2:
      token_endpoint: *TOKEN_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/oauth/token
      authorization_endpoint: *AUTHORIZATION_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/authorize
      credentials:
        client_id: *CLIENT_ID*
        client_secret: *CLIENT_SECRET*
      redirect_uri: /oauth2/callback
      redirect_path_matcher: /oauth2/callback
      signout_path: /oauth2/signout
      forward_bearer_token: true
      auth_scopes:
        - opened
```

## Non-Auth Policies: 

**Exposing OpenAPI API**
```yaml
# FROM 
x-kusk:
  openapi-path: openapi.json

# TO
x-kusk: 
  public_api_path: /openapi
```

**Disabled**: change to hidden, which reflects what it's actually doing
```yaml
# FROM 
/path:
  x-kusk:
    disabled: true
# TO 
/path:
  x-kusk:
    hidden: true
```

### PR DESCRIPTION
**Pull request by jasmingacic**
fixes #837 

This PR has breaking changes to the policies (formerly known as extensions) schema.

I decided to go with host instead of authorizer-hostname and authorizer-port. It makes it easier to copy past shorter property names
```
x-kusk:
  auth:
    cloudentity:
      host:
         hostname: cloudentity-authorizer-standalone-authorizer.kusk-system
         port: 9004
```
`scheme: Basic` is now `custom` 
```
x-kusk:
  auth:
    custom:
       host:
         hostname: example.com
         port: 80
```

```
x-kusk:
  auth:
    oauth2:
      token_endpoint: *TOKEN_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/oauth/token
      authorization_endpoint: *AUTHORIZATION_ENDPOINT* # <- for example https://yourdomain.eu.auth0.com/authorize
      credentials:
        client_id: *CLIENT_ID*
        client_secret: *CLIENT_SECRET*
      redirect_uri: /oauth2/callback
      redirect_path_matcher: /oauth2/callback
      signout_path: /oauth2/signout
      forward_bearer_token: true
      auth_scopes:
        - opened
```

This one is ugly but I opted for underscores instead of hyphen because it is easier to select words with underscore in it. I think we should take this property to the "drawing board"
```
x-kusk: 
  public_api_path: /openapi
```

`disabled` is now `hidden`. Hidden is more descriptive than disabled as the endpoint isn't really disabled but just hidden from the users.
```
/path:
  x-kusk:
    hidden: true
```

All tests have passed locally and on are remote cluster. We only need to wait on the CI to agree with that.

### CODE REVIEW

**Comment by jasmingacic** [comment]
While at this change we should start designing how JWT should fit into policies. I already have a proposal in this issue ->  https://github.com/kubeshop/kusk-gateway/issues/402

**Comment by jasmingacic** [comment]
Note: In this PR i have disabled check-cache as it was failing too often for integration tests to be practical

**Comment by mbana** [comment]
> Note: In this PR i have disabled check-cache as it was failing too often for integration tests to be practical

Good choice.

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ spec:
         - command:
             - /manager
           image: kusk-gateway:latest
-          imagePullPolicy: IfNotPresent
```
Comment (mbana) [inline config/manager/manager.yaml:29]: Why this change?

File: pkg/options/options.go [hunk 2]
```diff
@@ -111,31 +111,31 @@ func (o Options) Validate() error {
...
-	Auth        *AuthOptions       `json:"auth,omitempty" yaml:"auth,omitempty"`
+	Path          *PathOptions       `yaml:"path,omitempty" json:"path,omitempty"`
+	QoS           *QoSOptions        `yaml:"qos,omitempty" json:"qos,omitempty"`
+	CORS          *CORSOptions       `yaml:"cors,omitempty" json:"cors,omitempty"`
+	Websocket     *bool              `json:"websocket,omitempty" yaml:"websocket,omitempty"`
+	Validation    *ValidationOptions `json:"validation,omitempty" yaml:"validation,omitempty"`
+	Mocking       *MockingOptions    `json:"mocking,omitempty" yaml:"mocking,omitempty"`
+	RateLimit     *RateLimitOptions  `json:"rate_limit,omitempty" yaml:"rate_limit,omitempty"`
+	Cache         *CacheOptions      `json:"cache,omitempty" yaml:"cache,omitempty"`
+	PublicAPIPath string             `json:"public_api_path,omitempty" yaml:"public-api-path,omitempty"`
```
Comment (mbana) [inline pkg/options/options.go:128]: `public_api_path` vs `public-api-path`.

I would use `public-api-path`.

File: pkg/options/auth_test.go [hunk 3]
```diff
@@ -169,7 +162,8 @@ auth:
 			assert := assert.New(t)
 			options := &SubOptions{}
 			err := yaml.UnmarshalStrict([]byte(test.input), options)
-
+			jsn, _ := yaml.Marshal(test.expected)
+			fmt.Println(string(jsn))
```
Comment (mbana) [inline pkg/options/auth_test.go:166]: Please use `t.Log(...)` instead, or just remove it.

File: internal/controllers/parser.go [hunk 4]
```diff
@@ -441,12 +441,12 @@ func UpdateConfigFromAPIOpts(
...
 				perRouteAuth, err := auth.RouteAuthzDisabled()
 				if err != nil {
-					return fmt.Errorf("cannot create per-route config to disable authorization: openapi-path=%q, %w", opts.OpenAPIPath, err)
+					return fmt.Errorf("cannot create per-route config to disable authorization: public_api_path=%q, %w", opts.PublicAPIPath, err)
 				}
 
 				openapiRt.TypedPerFilterConfig[wellknown.HTTPExternalAuthorization] = perRouteAuth
 
-				logger.Info("disabled `auth` for route", "openapi-path", opts.OpenAPIPath, "vh", fmt.Sprintf("%q", string(vh)))
+				logger.Info("disabled `auth` for route", "public_api_path", opts.PublicAPIPath, "vh", fmt.Sprintf("%q", string(vh)))
```
Comment (mbana) [inline internal/controllers/parser.go:449]: Change to `public-api-path`, please.

File: internal/controllers/parser.go [hunk 5]
```diff
@@ -441,12 +441,12 @@ func UpdateConfigFromAPIOpts(
 
 				perRouteAuth, err := auth.RouteAuthzDisabled()
 				if err != nil {
-					return fmt.Errorf("cannot create per-route config to disable authorization: openapi-path=%q, %w", opts.OpenAPIPath, err)
+					return fmt.Errorf("cannot create per-route config to disable authorization: public_api_path=%q, %w", opts.PublicAPIPath, err)
```
Comment (mbana) [inline internal/controllers/parser.go:444]: Change to `public-api-path`, please.

File: Makefile [hunk 6]
```diff
@@ -159,7 +159,7 @@ docker-build: ## Build docker image with the manager.
 		--tag kubeshop/kusk-gateway:$(shell git describe --tags $(shell git rev-list --tags --max-count=1)) \
 		--file ./build/manager/Dockerfile \
 		.
-	minikube image --profile kgw load kubeshop/kusk-gateway:$(shell git describe --tags $(shell git rev-list --tags --max-count=1))
+	# minikube image --profile kgw load kubeshop/kusk-gateway:$(shell git describe --tags $(shell git rev-list --tags --max-count=1))
```
Comment (mbana) [inline Makefile:162]: This might effect @kylehodgetts.

I suggest you try out `make create-env.`

**Review by mbana — CHANGES_REQUESTED** [review]
LGTM.

Needs a few changes though.

File: pkg/options/auth_test.go (hunk 3 above)
Comment (jasmingacic) [reply pkg/options/auth_test.go:166]: I will have this removed. It doesn't serve any purpose here

File: pkg/options/options.go (hunk 2 above)
Comment (jasmingacic) [reply pkg/options/options.go:128]: underscore is so it is easer to mark the string when copying

File: config/manager/manager.yaml (hunk 1 above)
Comment (jasmingacic) [reply config/manager/manager.yaml]: reverted it

File: pkg/options/options.go (hunk 2 above)
Comment (jasmingacic) [reply pkg/options/options.go:128]: so we are going with `public_api_path` instead of `public-api-path` because the former is easier to select in the editor where words separated with hyphen require more than 3 clicks to select

File: internal/controllers/parser.go (hunk 5 above)
Comment (jasmingacic) [reply internal/controllers/parser.go:444]: see above please

File: Makefile (hunk 6 above)
Comment (jasmingacic) [reply Makefile]: done

File: internal/controllers/parser.go (hunk 4 above)
Comment (aabedraba) [reply internal/controllers/parser.go:449]: underscores are the pattern with all our extensions, that's why we decided to keep it common

File: docs/docs/extension.md [hunk 7]
```diff
@@ -366,8 +365,7 @@ Check the [Custom Upstream auth guide](./guides/authentication/custom-auth-upstr
 x-kusk:
 ...
   auth:
-    scheme: cloudentity
-    auth-upstream:
+    auth-cloudentity:
```
Comment (kylehodgetts) [inline docs/docs/extension.md:368]: Why not just `cloudentity` since it's already under the `auth` key?

File: pkg/options/auth_test.go [hunk 8]
```diff
@@ -42,28 +42,24 @@ func Test_AuthOptions_UnmarshalStrict(t *testing.T) {
...
 			expected: &AuthOptions{
-				Scheme:     "basic",
-				PathPrefix: StringToPtr("/login"),
-				AuthUpstream: &AuthUpstream{
+				Custom: &Custom{
+					PathPrefix: StringToPtr("/login"),
 					Host: AuthUpstreamHost{
 						Hostname: "example.com",
-						Port:     80,
+						// Port:     80,
```
Comment (kylehodgetts) [inline pkg/options/auth_test.go:55]: Remove if not needed

File: pkg/options/cache.go [hunk 9]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/cache.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/cors.go [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/cors.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/host.go [hunk 11]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/host.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/mocking.go [hunk 12]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/mocking.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/options.go [hunk 13]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/options.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/path.go [hunk 14]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/path.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/quos.go [hunk 15]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/quos.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: smoketests/samples/openapi-path/openapi-path-with-auth.yaml [hunk 16]
```diff
@@ -16,10 +16,9 @@ spec:
     - http
     - https
     x-kusk:
-      openapi-path: openapi.json
+      public_api_path: openapi.json
```
Comment (kylehodgetts) [inline smoketests/samples/openapi-path/openapi-path-with-auth.yaml:19]: Having underscores like this isn't common in YAML? Why have we opted for it?

File: pkg/options/ratelimit.go [hunk 17]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/ratelimit.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/redirect.go [hunk 18]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/redirect.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/rewrite.go [hunk 19]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/rewrite.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/static_options.go [hunk 20]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/static_options.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/upstream.go [hunk 21]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/upstream.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

File: pkg/options/validation.go [hunk 22]
```diff
@@ -1,7 +1,7 @@
 /*
 MIT License
 
-Copyright (c) 2022 Kubeshop
+# Copyright (c) 2022 Kubeshop
```
Comment (kylehodgetts) [inline pkg/options/validation.go:4]:
```suggestion
Copyright (c) 2022 Kubeshop
```

**Review by kylehodgetts — CHANGES_REQUESTED** [review]

File: docs/docs/extension.md (hunk 7 above)
Comment (jasmingacic) [reply docs/docs/extension.md:368]: I think I missed this

**Comment by jasmingacic** [comment]
@aabedraba do you need more input for the docs update? I tried to fix where I could

File: docs/docs/extension.md [hunk 23]
```diff
@@ -348,12 +348,11 @@ The `auth` object contains the following properties to configure HTTP authentica
...
-    path_prefix: /login # optional
-    auth-upstream:
-      host:
-        hostname: example.com
-        port: 80
+    custom:
+      path_prefix: /login # optional
+        host:
+          hostname: example.com
+          port: 80
```
Comment (aabedraba) [inline docs/docs/extension.md:355]:
```suggestion
      path_prefix: /login # optional
      host:
        hostname: example.com
        port: 80
```

File: pkg/options/auth.go [hunk 24]
```diff
@@ -48,57 +48,40 @@ const (
...
+		return fmt.Errorf("auth must have one of the following defined OAuth2, Custom or Cloudentity")
 	}
 
-	// `basic` or `cloudentity`
-	if SchemeBasic == o.Scheme || SchemeCloudEntity == o.Scheme {
-		return validation.ValidateStruct(&o,
-			validation.Field(&o.AuthUpstream, validation.Required),
-		)
+	if o.OAuth2 != nil && o.Custom != nil {
+		return fmt.Errorf("custom auth and OAuth2 cannot be enabled at the same time")
```
Comment (aabedraba) [inline pkg/options/auth.go:69]: I think we need a generic check: 
if: more than 1 auth mechanism is defined 
return: Only 1 authentication mechanism allowed

or something like that? 🤔

Comment (jasmingacic) [reply pkg/options/auth.go:69]: That's already in place
